### ISSUE TITLE
off-by-one bug in ProjectionActor when using akka-persistence-jdbc

### ISSUE THREAD

**Issue by rcavalcanti** [strongtyped/fun-cqrs#49]
Issue originaly reported by @vpavkin in Gitter. 
...
@rcavalcanti Hello again!
I've been playing around with FunCQRS, it is awesome! I have it working along with https://github.com/dnvriend/akka-persistence-jdbc plugin, and found a small issue I would like to hear your opinion on.
I'm sure there's some idea behind this implementation, so this is more of a "why it's the way it is" question than a bug-report :smile:
Code I talk about is this (https://github.com/strongtyped/fun-cqrs/blob/develop/modules/akka/src/main/scala/io/funcqrs/akka/ProjectionActor.scala#L70) :
sourceProvider.source(lastProcessedOffset.map(_ + 1).getOrElse(0)).runWith(actorSink)
The map here treats the offset sequence as 1-based, but the getOrElse clause as zero-based.
akka-persistence-jdbc plugin is inconsistent with that: it has 1-based offset. It works ok with 0, but when you restart the projection, it starts from the N + 1 and misses event N (here I talk about events that were persisted while the projection was off). Some adjustments needed to resolve this inconsistency.
Generally, the offset for eventsByTag is defined by each plugin and it's hard to expect an abstraction to support all the cases. But at least I would expect the abstraction to be consistent between the "first start" and "restart" cases.
Again, I might be missing something here.
Thanks for your help! :smile:
...

**Comment by rcavalcanti** [comment]
Hi @vpavkin,

After some research, I have the impression that the error is on the plugin side. 

I see the following on their source:

```scala
  def eventsByTag(tag: String, offset: Long): Query[Journal, JournalRow, Seq] =
    JournalTable.filter(_.tags like s"%$tag%").sortBy(_.created.asc).drop(offset)

  def eventsByTagAndPersistenceId(persistenceId: String, tag: String, offset: Long): Query[Journal, JournalRow, Seq] =
    JournalTable.filter(_.persistenceId === persistenceId).filter(_.tags like s"%$tag%").sortBy(_.sequenceNumber.asc).filter(_.sequenceNumber >= offset)
```

In `eventsByTag`, they drop everything up to the offset (offset included).

In `eventsByTagAndPersistenceId`, they retain everything with `sequenceNumber >= offset`, thus offset included in the result.

The second is aligned with LevelDB and Cassandra plugin (and with our internal Postgres plugin as well).

The problem is that the query documentation is lacking and we don't have a TestKit yet. 

Maybe @dnvriend can shed some light here.

**Comment by vpavkin** [comment]
@rcavalcanti Sorry for being silent for some time. I see what you mean, definitely makes sense.
I agree that the main problem here is the lack of precise persistent query spec. Thanks for digging the truth! :)

Luckily, a workaround here is rather simple. Feel free to close this issue. :+1:

**Comment by rcavalcanti** [comment]
Hi @vpavkin,

I won't close this issue for the moment as I want to investigate better what are the plans for the query API. I want to investigate it further before reaching the Akka team.

**Comment by vpavkin** [comment]
hi @rcavalcanti.
Just upgraded my project to FunCQRS 0.4.6 and akka-persistence-jdbc 2.5.2.
Bug doesn't seem to reproduce any more.

**Comment by rcavalcanti** [comment]
Hi @vpavkin,

Nothing was added to on Fun.CQRS with respect to that. I think the fix comes from akka-persistence-jdbc 2.5.2.

I have saw some tweets from @dnvriend with the akka team discussing about the TestKit for akka persistence query. He may get involved with it. 

I'll close this issue therefore.
